// electron-builder-lib/out/targets/targetUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.createHelperDir = exports.StageDir = undefined;

var _bluebirdLst;

function _load_bluebirdLst() {
    return _bluebirdLst = require("bluebird-lst");
}

var _bluebirdLst2;

function _load_bluebirdLst2() {
    return _bluebirdLst2 = _interopRequireDefault(require("bluebird-lst"));
}

let createHelperDir = exports.createHelperDir = (() => {
    var _ref = (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* (target, arch) {
        const tempDir = _path.join(target.outDir, `__${target.name}-temp-${(_builderUtil || _load_builderUtil()).Arch[arch]}`);
        yield (0, (_fsExtraP || _load_fsExtraP()).emptyDir)(tempDir);
        return new StageDir(tempDir);
    });

    return function createHelperDir(_x, _x2) {
        return _ref.apply(this, arguments);
    };
})();


var _fsExtraP;

function _load_fsExtraP() {
    return _fsExtraP = require("fs-extra-p");
}

var _path = _interopRequireWildcard(require("path"));

var _builderUtil;

function _load_builderUtil() {
    return _builderUtil = require("builder-util");
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class StageDir {
    constructor(tempDir) {
        this.tempDir = tempDir;
    }
    ensureEmpty() {
        return (0, (_fsExtraP || _load_fsExtraP()).emptyDir)(this.tempDir);
    }
    getTempFile(name) {
        return this.tempDir + _path.sep + name;
    }
    cleanup() {
        if (!(_builderUtil || _load_builderUtil()).debug.enabled || process.env.ELECTRON_BUILDER_REMOVE_STAGE_EVEN_IF_DEBUG === "true") {
            return (0, (_fsExtraP || _load_fsExtraP()).remove)(this.tempDir);
        }
        return (_bluebirdLst2 || _load_bluebirdLst2()).default.resolve();
    }
}
exports.StageDir = StageDir;
